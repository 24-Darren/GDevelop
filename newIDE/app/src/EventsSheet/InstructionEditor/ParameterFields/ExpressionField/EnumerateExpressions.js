// @flow weak
import update from 'lodash.update';
import compact from 'lodash.compact';
const gd = global.gd;

const GROUP_DELIMITER = '/';

const enumerateExtensionExpressions = (
  prefix,
  expressions,
  objectMetadata,
  behaviorMetadata
) => {
  const allExpressions = [];

  //Get the map containing the metadata of the expression provided by the extension...
  var expressionsTypes = expressions.keys();

  //... and add each instruction
  for (var j = 0; j < expressionsTypes.size(); ++j) {
    var exprMetadata = expressions.get(expressionsTypes.get(j));
    if (!exprMetadata.isShown()) {
      //Skip hidden expressions
      continue;
    }

    var parameters = [];
    for (var i = 0; i < exprMetadata.getParametersCount(); i++) {
      if (objectMetadata && i == 0) continue;
      if (behaviorMetadata && i <= 1) continue; //Skip object and behavior parameters
      if (exprMetadata.getParameter(i).isCodeOnly()) continue;

      parameters.push(exprMetadata.getParameter(i));
    }

    const displayedName = exprMetadata.getFullName();
    const groupName = exprMetadata.getGroup();
    const fullGroupName = prefix + groupName;

    allExpressions.push({
      type: expressionsTypes.get(j),
      displayedName,
      fullGroupName,
      metadata: exprMetadata,
      parameters: parameters,
      objectMetadata: objectMetadata,
      behaviorMetadata: behaviorMetadata,
    });
  }

  return allExpressions;
};

export const enumerateExpressions = (type = 'number') => {
  const freeExpressions = [];
  const objectsExpressions = [];
  const behaviorsExpressions = [];

  const allExtensions = gd
    .asPlatform(gd.JsPlatform.get())
    .getAllPlatformExtensions();
  for (var i = 0; i < allExtensions.size(); ++i) {
    var extension = allExtensions.get(i);
    var allObjectsTypes = extension.getExtensionObjectsTypes();
    var allBehaviorsTypes = extension.getBehaviorsTypes();

    //TODO: Factor with InstructionTypeSelector.js
    let prefix = '';
    if (allObjectsTypes.size() > 0 || allBehaviorsTypes.size() > 0) {
      prefix =
        extension.getName() === 'BuiltinObject'
          ? 'Common expressions for all objects'
          : extension.getFullName();
      prefix += GROUP_DELIMITER;
    }

    //Check which type of expression we want to autocomplete.
    var allFreeExpressionsGetter = extension.getAllExpressions;
    var allObjectExpressionsGetter = extension.getAllExpressionsForObject;
    var allBehaviorExpressionsGetter = extension.getAllExpressionsForBehavior;
    if (type === 'string') {
      allFreeExpressionsGetter = extension.getAllStrExpressions;
      allObjectExpressionsGetter = extension.getAllStrExpressionsForObject;
      allBehaviorExpressionsGetter = extension.getAllStrExpressionsForBehavior;
    }

    //Free expressions
    freeExpressions.push.apply(
      freeExpressions,
      enumerateExtensionExpressions(prefix, allFreeExpressionsGetter.call(extension))
    );

    //Objects expressions:
    for (var j = 0; j < allObjectsTypes.size(); ++j) {
      var objMetadata = extension.getObjectMetadata(allObjectsTypes.get(j));
      objectsExpressions.push.apply(
        objectsExpressions,
        enumerateExtensionExpressions(
          prefix,
          allObjectExpressionsGetter.call(extension, allObjectsTypes.get(j)),
          objMetadata
        )
      );
    }

    //Behaviors expressions:
    for (var j = 0; j < allBehaviorsTypes.size(); ++j) {
      var autoMetadata = extension.getBehaviorMetadata(
        allBehaviorsTypes.get(j)
      );
      behaviorsExpressions.push.apply(
        behaviorsExpressions,
        enumerateExtensionExpressions(
          prefix,
          allBehaviorExpressionsGetter.call(
            extension,
            allBehaviorsTypes.get(j)
          ),
          undefined,
          autoMetadata
        )
      );
    }
  }

  return {
    allExpressions: [
      ...freeExpressions,
      ...objectsExpressions,
      ...behaviorsExpressions,
    ],
    freeExpressions,
    objectsExpressions,
    behaviorsExpressions,
  };
};

export const filterExpressions = (list, searchText) => {
    if (!searchText) return list;
    const lowercaseSearchText = searchText.toLowerCase();

    return list.filter(enumeratedExpression => {
      return (
        enumeratedExpression.type
          .toLowerCase()
          .indexOf(lowercaseSearchText) !== -1
      );
    });
}

export const createExpressionsTree = (allExpressions) => {
  const tree = {};
  allExpressions.forEach(expressionInfo => {
    update(
      tree,
      compact(expressionInfo.fullGroupName.split(GROUP_DELIMITER)),
      groupInfo => {
        const existingGroupInfo = groupInfo || {};
        return {
          ...existingGroupInfo,
          [expressionInfo.displayedName]: expressionInfo,
        };
      }
    );
  });

  return tree;
}
