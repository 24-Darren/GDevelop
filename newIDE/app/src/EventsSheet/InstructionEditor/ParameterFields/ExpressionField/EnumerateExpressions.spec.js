import { enumerateExpressions, filterExpressions, createExpressionsTree } from './EnumerateExpressions';

describe('EnumerateObjects', () => {
  it('can enumerate and filter expressions', () => {
    const {
      freeExpressions,
      objectsExpressions,
    } = enumerateExpressions('number');

    // Should find atan, atan2, atanh math function
    expect(filterExpressions(freeExpressions, 'atan')).toHaveLength(3);

    // Should find abs math function
    expect(filterExpressions(freeExpressions, 'abs')).toHaveLength(1);

    expect(filterExpressions(freeExpressions, 'MouseX')).toHaveLength(1);
    expect(filterExpressions(freeExpressions, 'MouseY')).toHaveLength(1);

    expect(filterExpressions(objectsExpressions, 'PointX')).toHaveLength(1);
  });

  it('can create the tree of some instructions', () => {
    const {
      objectsExpressions,
    } = enumerateExpressions('number');
    expect(createExpressionsTree(objectsExpressions)).toMatchSnapshot();
  })
});
