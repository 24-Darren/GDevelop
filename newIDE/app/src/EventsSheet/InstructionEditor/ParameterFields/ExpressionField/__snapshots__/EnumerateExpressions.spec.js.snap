// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnumerateObjects can create the tree of some instructions 1`] = `
Object {
  "Common expressions for all objects": Object {
    "Angle": Object {
      "Angle": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Angle",
        "fullGroupName": "Common expressions for all objects/Angle",
        "metadata": ExpressionMetadata {
          "ptr": 5954072,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "Angle",
      },
    },
    "Movement": Object {
      "Average X coordinates of forces": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Average X coordinates of forces",
        "fullGroupName": "Common expressions for all objects/Movement",
        "metadata": ExpressionMetadata {
          "ptr": 5954328,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "ForceX",
      },
      "Average Y coordinates of forces": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Average Y coordinates of forces",
        "fullGroupName": "Common expressions for all objects/Movement",
        "metadata": ExpressionMetadata {
          "ptr": 5954664,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "ForceY",
      },
      "Average angle of the forces": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Average angle of the forces",
        "fullGroupName": "Common expressions for all objects/Movement",
        "metadata": ExpressionMetadata {
          "ptr": 5955040,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "ForceAngle",
      },
      "Average length of the forces": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Average length of the forces",
        "fullGroupName": "Common expressions for all objects/Movement",
        "metadata": ExpressionMetadata {
          "ptr": 5955416,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "ForceLength",
      },
    },
    "Position": Object {
      "Distance between two objects": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Distance between two objects",
        "fullGroupName": "Common expressions for all objects/Position",
        "metadata": ExpressionMetadata {
          "ptr": 5958184,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 5958524,
          },
        ],
        "type": "Distance",
      },
      "Square distance between two objects": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Square distance between two objects",
        "fullGroupName": "Common expressions for all objects/Position",
        "metadata": ExpressionMetadata {
          "ptr": 5958792,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 5959124,
          },
        ],
        "type": "SqDistance",
      },
      "X position": Object {
        "behaviorMetadata": undefined,
        "displayedName": "X position",
        "fullGroupName": "Common expressions for all objects/Position",
        "metadata": ExpressionMetadata {
          "ptr": 5953688,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "X",
      },
      "Y position": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Y position",
        "fullGroupName": "Common expressions for all objects/Position",
        "metadata": ExpressionMetadata {
          "ptr": 5953880,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "Y",
      },
    },
    "Size": Object {
      "Height": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Height",
        "fullGroupName": "Common expressions for all objects/Size",
        "metadata": ExpressionMetadata {
          "ptr": 5956760,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "Height",
      },
      "Width": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Width",
        "fullGroupName": "Common expressions for all objects/Size",
        "metadata": ExpressionMetadata {
          "ptr": 5956168,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "Width",
      },
    },
    "Variables": Object {
      "Object's variable": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Object's variable",
        "fullGroupName": "Common expressions for all objects/Variables",
        "metadata": ExpressionMetadata {
          "ptr": 5959224,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 5959444,
          },
        ],
        "type": "Variable",
      },
      "Object's variable number of children": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Object's variable number of children",
        "fullGroupName": "Common expressions for all objects/Variables",
        "metadata": ExpressionMetadata {
          "ptr": 5959656,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 5960068,
          },
        ],
        "type": "VariableChildCount",
      },
    },
    "Visibility": Object {
      "Z order": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Z order",
        "fullGroupName": "Common expressions for all objects/Visibility",
        "metadata": ExpressionMetadata {
          "ptr": 5957432,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5913668,
        },
        "parameters": Array [],
        "type": "ZOrder",
      },
    },
  },
  "Sprite": Object {
    "Animations and images": Object {
      "Animation": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Animation",
        "fullGroupName": "Sprite/Animations and images",
        "metadata": ExpressionMetadata {
          "ptr": 6004352,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "Animation",
      },
      "Animation speed scale": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Animation speed scale",
        "fullGroupName": "Sprite/Animations and images",
        "metadata": ExpressionMetadata {
          "ptr": 6005680,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "AnimationSpeedScale",
      },
      "Image": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Image",
        "fullGroupName": "Sprite/Animations and images",
        "metadata": ExpressionMetadata {
          "ptr": 6005208,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "Sprite",
      },
    },
    "Direction": Object {
      "Direction": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Direction",
        "fullGroupName": "Sprite/Direction",
        "metadata": ExpressionMetadata {
          "ptr": 6003544,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "Direction",
      },
    },
    "Position": Object {
      "X position of a point": Object {
        "behaviorMetadata": undefined,
        "displayedName": "X position of a point",
        "fullGroupName": "Sprite/Position",
        "metadata": ExpressionMetadata {
          "ptr": 6002328,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 6002668,
          },
        ],
        "type": "PointX",
      },
      "Y position of a point": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Y position of a point",
        "fullGroupName": "Sprite/Position",
        "metadata": ExpressionMetadata {
          "ptr": 6002808,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [
          ParameterMetadata {
            "ptr": 6003148,
          },
        ],
        "type": "PointY",
      },
    },
    "Size": Object {
      "Scale of the height of an object": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Scale of the height of an object",
        "fullGroupName": "Sprite/Size",
        "metadata": ExpressionMetadata {
          "ptr": 6006632,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "ScaleY",
      },
      "Scale of the width of an object": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Scale of the width of an object",
        "fullGroupName": "Sprite/Size",
        "metadata": ExpressionMetadata {
          "ptr": 6006072,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 5950172,
        },
        "parameters": Array [],
        "type": "ScaleX",
      },
    },
  },
  "Text object": Object {
    "Opacity": Object {
      "Opacity": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Opacity",
        "fullGroupName": "Text object/Opacity",
        "metadata": ExpressionMetadata {
          "ptr": 6299080,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 6283324,
        },
        "parameters": Array [],
        "type": "Opacity",
      },
    },
    "Rotation": Object {
      "Angle": Object {
        "behaviorMetadata": undefined,
        "displayedName": "Angle",
        "fullGroupName": "Text object/Rotation",
        "metadata": ExpressionMetadata {
          "ptr": 6299536,
        },
        "objectMetadata": ObjectMetadata {
          "ptr": 6283324,
        },
        "parameters": Array [],
        "type": "Angle",
      },
    },
  },
}
`;
